Add missing metadata

tfrommen opened the pull request:
This PR adds missing metadata, and it ensures that metadata gets actually saved.

johnbillion reviewed (changes requested):
It looks like these should go into the `Playable` class because they're supported by videos too. Ref: https://github.com/WordPress/wordpress-develop/blob/24f0d03a6c6b4a21da12cc4b7f1179adec155312/src/wp-admin/includes/media.php#L3380-L3384

In addition, there's a bunch of other metadata fields available that are read by `wp_read_video_metadata()` and `wp_read_audio_metadata()`. I think the `Media` class needs `set_metadata( array $meta )` and `add_metadata( $key, $value )` methods to set these metadata fields, and use those instead of the proposed `set_bitrate()` here.

I think this might also apply to `set_album()` and `set_artist()`, it looks like the media template checks for `data.album` in a few places but always falls back to `data.meta.album`, and the same for `artist`. Ref: https://github.com/WordPress/wordpress-develop/blob/24f0d03a6c6b4a21da12cc4b7f1179adec155312/src/wp-includes/media-template.php#L490-L499 .

Thoughts?

tfrommen commented:
> It looks like these should go into the `Playable` class because they're supported by videos too. Ref: https://github.com/WordPress/wordpress-develop/blob/24f0d03a6c6b4a21da12cc4b7f1179adec155312/src/wp-admin/includes/media.php#L3380-L3384

Hm, OK, that's interesting. I based my PRs addressing missing data/setters onwhat I could find in the media templates. And all places that use artis, album and bitrate are audio-only:

- https://github.com/WordPress/wordpress-develop/blob/24f0d03a6c6b4a21da12cc4b7f1179adec155312/src/wp-includes/media-template.php#L453-L460
- https://github.com/WordPress/wordpress-develop/blob/24f0d03a6c6b4a21da12cc4b7f1179adec155312/src/wp-includes/media-template.php#L488-L500
- https://github.com/WordPress/wordpress-develop/blob/24f0d03a6c6b4a21da12cc4b7f1179adec155312/src/wp-includes/media-template.php#L717-L729

The `Playable` class currently already has `set_meta`, which is for the entire data (array). We could move that to `Media`. However, adding another method `set_metadata` might be confusing. Either it should be named something like `set_meta_field`, or the different classes should provide dedicated setters for certain fields (or multiple). This would also allow for defining/enforcing types, which I think would be good to have for select fields. Of course there is more arbitrary metadata that could get added, and that is then freeform.

Now, looking at WordPress again, all usage of `artist` and `album` is for audio only, also when uploading a file:

- https://github.com/WordPress/wordpress-develop/blob/24f0d03a6c6b4a21da12cc4b7f1179adec155312/src/wp-admin/includes/media.php#L320-L354

So I feel these should get added to `Audio` and not `Playable`.

`bitrate` and `bitrate_mode`, on the other hand, may be present for both audio and video, so I agree to move it to `Playable` instead.

For now, I think I will leave this open and not include in the upcoming release, and come back to this some time soon.

tfrommen commented:
@johnbillion I just updated this PR.

- Moved the `set_meta` setter from `Playable` to `Media`, and added a new `add_meta` method for setting a specific meta field (similar to `add_amf_meta`).
- Moved the `set_bitrate` method from `Audio` to `Playbale`, as discussed earlier.
- Used the new `add_meta` method for `album`, `artist` and `bitrate(_mode)`.
- **Important:** Ensured that data provided in the `meta` property gets actually saved!
- All optional (typeless) metadata fields can then be set using the `add_meta` method.
- In the AMF WordPress plugin/provider, I will then store all data in `media_details` under the `_wp_attachment_metadata` meta key.

Does that make sense to you?
